File: src/domain/network.py
"""Pure network requirement policies."""

from __future__ import annotations

import ipaddress
from typing import Any, Dict, Mapping, Optional, Tuple


_VPN_FLAG = "k3s_use_socks5_proxy"
_VPN_FLAG_LEGACY = "argocd_use_socks5_proxy"
_ANSIBLE_HOST = "ansible_host"


def _extract_group_vars(inv_data: Any, group_name: str) -> Mapping[str, Any]:
    if not isinstance(inv_data, dict):
        return {}

    all_data = inv_data.get("all")
    if not isinstance(all_data, dict):
        return {}

    children = all_data.get("children")
    if not isinstance(children, dict):
        return {}

    group_data = children.get(group_name)
    if not isinstance(group_data, dict):
        return {}

    vars_dict = group_data.get("vars")
    if not isinstance(vars_dict, Mapping):
        return {}

    return vars_dict


def _network_range_for_host(ansible_host: str) -> tuple[Optional[str], Optional[str]]:
    if not is_private_network(ansible_host):
        return None, None

    try:
        ip = ipaddress.ip_address(ansible_host)
        network = ipaddress.ip_network(f"{ip}/24", strict=False)
        return "sshuttle", str(network)
    except ValueError:
        return "network", None


def is_private_network(ip_or_hostname: str) -> bool:
    try:
        ip = ipaddress.ip_address(ip_or_hostname)
        return ip.is_private
    except ValueError:
        return False


def check_vpn_requirement(inv_data: Any, group_name: str, host_name: str) -> bool:
    del host_name
    vars_dict = _extract_group_vars(inv_data, group_name)
    return bool(vars_dict.get(_VPN_FLAG, False)) or bool(vars_dict.get(_VPN_FLAG_LEGACY, False))


def check_network_requirement(hostname: str, host_info: Dict[str, Any]) -> Tuple[Optional[str], Optional[str]]:
    del hostname
    config = host_info.get("config", {})
    ansible_host = config.get(_ANSIBLE_HOST)

    if not ansible_host:
        return None, None

    return _network_range_for_host(str(ansible_host))


def detect_network_requirement(
    host_config: Mapping[str, Any] | Dict[str, Any],
    group_vars: Mapping[str, Any] | Dict[str, Any] | None = None,
    group_name: Optional[str] = None,
) -> tuple[Optional[str], Optional[str], bool]:
    host_info = {"config": dict(host_config)}
    network_type, network_range = check_network_requirement(group_name or "", host_info)

    needs_vpn = bool(host_config.get(_VPN_FLAG, False)) or bool(host_config.get(_VPN_FLAG_LEGACY, False))
    vars_dict = host_config.get("vars")
    if isinstance(vars_dict, Mapping):
        needs_vpn = needs_vpn or bool(vars_dict.get(_VPN_FLAG, False)) or bool(vars_dict.get(_VPN_FLAG_LEGACY, False))
    if isinstance(group_vars, Mapping):
        needs_vpn = needs_vpn or bool(group_vars.get(_VPN_FLAG, False)) or bool(group_vars.get(_VPN_FLAG_LEGACY, False))

    return network_type, network_range, needs_vpn

File: internal/domain/network.go
package domain

import (
	"fmt"
	"net"
)

const (
	vpnFlag       = "k3s_use_socks5_proxy"
	vpnFlagLegacy = "argocd_use_socks5_proxy"
	hostAddr      = "addr"
)

// IsPrivateNetwork returns true if ip is an RFC 1918 private address.
func IsPrivateNetwork(ip string) bool {
	parsed := net.ParseIP(ip)
	if parsed == nil {
		return false
	}
	private := []string{
		"10.0.0.0/8",
		"172.16.0.0/12",
		"192.168.0.0/16",
	}
	for _, cidr := range private {
		_, network, _ := net.ParseCIDR(cidr)
		if network.Contains(parsed) {
			return true
		}
	}
	return false
}

// CheckVPNRequirement returns true when the inventory group declares a VPN/socks5 proxy requirement.
func CheckVPNRequirement(invData map[string]any, groupName, _ string) bool {
	vars := extractGroupVars(invData, groupName)
	return toBool(vars[vpnFlag]) || toBool(vars[vpnFlagLegacy])
}

// CheckNetworkRequirement returns (networkType, networkRange) for a host_info map.
// host_info must have a "config" key containing "addr".
func CheckNetworkRequirement(_ string, hostInfo map[string]any) (string, string) {
	config, _ := hostInfo["config"].(map[string]any)
	if config == nil {
		return "", ""
	}
	ah, _ := config[hostAddr].(string)
	if ah == "" {
		return "", ""
	}
	return networkRangeForHost(ah)
}

// DetectNetworkRequirement derives network type, range, and VPN flag from host config + group vars.
func DetectNetworkRequirement(hostConfig, groupVars map[string]any, _ string) (string, string, bool) {
	hostInfo := map[string]any{"config": hostConfig}
	netType, netRange := CheckNetworkRequirement("", hostInfo)

	needsVPN := toBool(hostConfig[vpnFlag]) || toBool(hostConfig[vpnFlagLegacy])
	if vars, ok := hostConfig["vars"].(map[string]any); ok {
		needsVPN = needsVPN || toBool(vars[vpnFlag]) || toBool(vars[vpnFlagLegacy])
	}
	if groupVars != nil {
		needsVPN = needsVPN || toBool(groupVars[vpnFlag]) || toBool(groupVars[vpnFlagLegacy])
	}

	return netType, netRange, needsVPN
}

func networkRangeForHost(ip string) (string, string) {
	if !IsPrivateNetwork(ip) {
		return "", ""
	}
	parsed := net.ParseIP(ip)
	if parsed == nil {
		return "network", ""
	}
	_, network, err := net.ParseCIDR(fmt.Sprintf("%s/24", parsed.String()))
	if err != nil {
		return "network", ""
	}
	return "sshuttle", network.String()
}

func extractGroupVars(invData map[string]any, groupName string) map[string]any {
	if invData == nil {
		return nil
	}
	all, _ := invData["all"].(map[string]any)
	children, _ := all["children"].(map[string]any)
	group, _ := children[groupName].(map[string]any)
	vars, _ := group["vars"].(map[string]any)
	return vars
}
